### DISCUSSION TITLE
cannot import name 'ExecutionContext' from 'sqlalchemy' when trying to open an MS-Access DB

### DISCUSSION THREAD

**Discussion by BeatArnet** [sqlalchemy/sqlalchemy#9727] [Usage Questions] [answered]
Programming environment:
- Python 3.10
- JupyterLab
- Windows 11

I am trying to open a local accdb database:

``
import sqlalchemy as sa

connection_string = (
    r "DRIVER={Microsoft Access Driver (*.mdb, *.accdb)};"
    r "DBQ=C:\Users\test\test.accdb;"
    r "ExtendedAnsiSQL=1;"
    r "MODE=Read;"
    )

connection_url = sa.engine.URL.create(
    "access+pyodbc",
    query={"odbc_connect": connection_string}
    )

print (connection_string)
engine = sa.create_engine(connection_url)
``
and get the following error message:
ImportError: cannot import name 'ExecutionContext' from 'sqlalchemy' (C:\Users\beata\anaconda3\lib\site-packages\sqlalchemy\__init__.py)

The same error occurs when I call the database using the assigned DSN (created with 64-bit ODBC Data Sources).

**Comment by zzzeek** [comment, marked as answer]
need to make sure all software is correctly installed, cc @gordthompson who maintains the access dialect.   

can we see the full stack trace for what's being imported where?   also if you can report on:

1. version of SQLAlchemy installed
2. version of sqlalchemy-access installed

**Reply by BeatArnet** [reply]
Hi Mike

Thanks for the hint

After I updated both libraries to the latest versions, it worked then almost without problems.

**pip install sqlalchemy**
_Requirement already satisfied: sqlalchemy in c:\users\beata\anaconda3\lib\site-packages (2.0.11)
Requirement already satisfied: greenlet!=0.4.17 in c:\users\beata\anaconda3\lib\site-packages (from sqlalchemy) (1.1.1)
Requirement already satisfied: typing-extensions>=4.2.0 in c:\users\beata\anaconda3\lib\site-packages (from sqlalchemy) (4.5.0)
Note: you may need to restart the kernel to use updated packages._

**pip install sqlalchemy-access**
_Requirement already satisfied: sqlalchemy-access in c:\users\beata\anaconda3\lib\site-packages (2.0.1)
Requirement already satisfied: SQLAlchemy>=2.0.0 in c:\users\beata\anaconda3\lib\site-packages (from sqlalchemy-access) (2.0.11)
Requirement already satisfied: pywin32 in c:\users\beata\anaconda3\lib\site-packages (from sqlalchemy-access) (306)
Requirement already satisfied: pyodbc>=4.0.27 in c:\users\beata\anaconda3\lib\site-packages (from sqlalchemy-access) (4.0.34)
Requirement already satisfied: greenlet!=0.4.17 in c:\users\beata\anaconda3\lib\site-packages (from SQLAlchemy>=2.0.0->sqlalchemy-access) (1.1.1)
Requirement already satisfied: typing-extensions>=4.2.0 in c:\users\beata\anaconda3\lib\site-packages (from SQLAlchemy>=2.0.0->sqlalchemy-access) (4.5.0)
Note: you may need to restart the kernel to use updated packages._

I just had to add the **text()** function to the pd.read_sql() function:

```py
df = pd.read_sql(text(query), conn)
````

Beat

**Reply by gordthompson** [reply]
> I just had to add the text() function to the pd.read_sql() function

Upgrading pandas to the latest version should avoid that issue.

**Comment by BeatArnet** [comment]
That helped, problem solved.

Thanks a lot

 

Von: Gord Thompson ***@***.***> 
Gesendet: Sonntag, 30. April 2023 20:35
An: sqlalchemy/sqlalchemy ***@***.***>
Cc: Beat Arnet ***@***.***>; Author ***@***.***>
Betreff: Re: [sqlalchemy/sqlalchemy] cannot import name 'ExecutionContext' from 'sqlalchemy' when trying to open an MS-Access DB (Discussion #9727)

 

I just had to add the text() function to the pd.read_sql() function

Upgrading pandas to the latest version should avoid that issue.
